### PR TITLE
Adding Code Coverage webserver code and Fix TypeError and AttributeError in Views.py

### PR DESCRIPTION
**Pull request by andyjianzhou**
Improved code coverage in these following spaces in both Webserver and Core:

Run: **Core**: `python scripts/cov/core_coverage.py`
Run: **Webserver**: `python scripts/cov/www_coverage.py`

`airflow/models/taskmap.py` - Increased from 82% to 100% Coverage
`airflow/www/widgets.py` - Increased from 79% to 100% Coverage
`airflow/www/views.py` - Help increase Coverage from 78% to 85%
Related: #35127 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/www/views.py [hunk 1]
```diff
@@ -271,7 +271,12 @@ def get_date_time_num_runs_dag_runs_form_data(www_request, session, dag):
         base_date = (date_time + datetime.timedelta(seconds=1)).replace(microsecond=0)
 
     default_dag_run = conf.getint("webserver", "default_dag_run_display_number")
-    num_runs = www_request.args.get("num_runs", default=default_dag_run, type=int)
```
Comment (aritra24) [inline airflow/www/views.py]: Why replace this? Both seem to be performing the same operation?

File: tests/www/test_validators.py [hunk 2]
```diff
@@ -164,4 +162,4 @@ def setup_method(self):
     def test_read_only_validator(self):
         validator = validators.ReadOnly()
         assert validator(self.form_mock, self.form_read_only_field_mock) is None
-        assert self.form_read_only_field_mock.flags.readonly is True
```
Comment (aritra24) [inline tests/www/test_validators.py]: This seems to be something maybe done by the ide, might be good to fix it.

File: tests/www/views/test_views_grid.py [hunk 3]
```diff
@@ -514,3 +515,59 @@ def test_next_run_datasets_404(admin_client):
...
+    })
+
+    with create_session() as session:
+        data = get_date_time_num_runs_dag_runs_form_data(request_with_base_date, session, run1.dag)
+
+    assert data['base_date'] == _safe_parse_datetime(base_date)
+    assert data['execution_date'] == run1.execution_date.isoformat()
+    assert data['num_runs'] == 5
+
+    # Test case 3: both run_id and base_date are provided
```
Comment (aritra24) [inline tests/www/views/test_views_grid.py]: I'm not sure if test case 1 & 2 are testing anything new that's not also done in 3? If they aren't maybe it might be better to skip them and just have 3.

Comment (andyjianzhou) [reply tests/www/views/test_views_grid.py]: Sounds good!

File: tests/www/test_validators.py (hunk 2 above)
Comment (andyjianzhou) [reply tests/www/test_validators.py]: Hmm weird, I think this is something done maybe by my IDE. Let me check it really quickly

Comment (andyjianzhou) [reply tests/www/test_validators.py]: Oh? Wait, I think that testcase was already there to begin with. What needs fixing for it?
`assert self.form_read_only_field_mock.flags.readonly is True`

File: tests/www/views/test_views_grid.py (hunk 3 above)
Comment (andyjianzhou) [reply tests/www/views/test_views_grid.py]: You're definitely right, a lot of redundant code. I made sure to remove that

File: airflow/www/views.py (hunk 1 above)
Comment (andyjianzhou) [reply airflow/www/views.py]: The change was made to ensure compatibility across different environments and setups, particularly in testing scenarios where request.args might not support keyword arguments. For me, I obtained a Keyword error while trying to create my mock data. Although this does the same operation, some formdata does not support keyword arguments. Correct me if I'm wrong tho :D

File: tests/www/views/test_views_grid.py (hunk 3 above)
Comment (andyjianzhou) [reply tests/www/views/test_views_grid.py]: I'll also be adding more tests for other places around the code as well. Ryan recommended testing in Core, so I might shift my focus there after I hammer out a few more unittests on the webserver.

File: airflow/www/views.py (hunk 1 above)
Comment (aritra24) [reply airflow/www/views.py]: Ideally you don't want to change the code to fit the test, you'd be better off changing the test to fit the code. I saw in the test you were creating a custom class, it might be better to use the class that's actually passed into the method. Perhaps some of the other tests might be using it, that should help you out.

File: tests/www/test_validators.py (hunk 2 above)
Comment (aritra24) [reply tests/www/test_validators.py]: It's the new line at the end that got removed I believe

File: airflow/www/views.py (hunk 1 above)
Comment (andyjianzhou) [reply airflow/www/views.py]: I see, how can I find the class that was passed into the method? I tried to replicate the class that was passed into the class. It seems like it was being passed through another function via API? Thanks!

File: tests/www/test_validators.py (hunk 2 above)
Comment (andyjianzhou) [reply tests/www/test_validators.py]: Just added it

File: airflow/www/views.py (hunk 1 above)
Comment (andyjianzhou) [reply airflow/www/views.py]: I think I got it figured out. We can instead use response to pass it in as it's used via API call.

Comment (andyjianzhou) [reply airflow/www/views.py]: Got it fixed in my other commit. Boosted code coverage to 85%

**Comment by uranusjr** [comment]
You’ll need to revert the whitespace changes, they cause the linter to fail.

**Comment by andyjianzhou** [comment]
> You’ll need to revert the whitespace changes, they cause the linter to fail.

Will do thanks!

**Review by RNHTTR — APPROVED** [review]

File: tests/www/views/test_views_tasks.py [hunk 4]
```diff
@@ -66,7 +66,7 @@ def reset_dagruns():
 
 
 @pytest.fixture(autouse=True)
-def init_dagruns(app, reset_dagruns):
```
Comment (aritra24) [inline tests/www/views/test_views_tasks.py:69]: What was this used for?

File: tests/models/test_xcom_arg_map.py [hunk 5]
```diff
@@ -188,6 +191,44 @@ def does_not_work_with_c(v):
     ]
 
 
+def test_task_map_from_task_instance_xcom():
```
Comment (aritra24) [inline tests/models/test_xcom_arg_map.py:194]: Mostly for my understanding how is this related to xcom? Also perhaps the two cases should be two different test cases one a success and one fail but mostly a nit.

**Review by aritra24 — APPROVED** [review]
Looks fine to me, a few nits and questions

File: tests/www/views/test_views_tasks.py (hunk 4 above)
Comment (andyjianzhou) [reply tests/www/views/test_views_tasks.py:69]: This basically initializes the DAG runs with pre-made execution dates and specific states. Basically for every test, it runs for every test in the module, setting states to RUNNING

File: tests/models/test_xcom_arg_map.py (hunk 5 above)
Comment (andyjianzhou) [reply tests/models/test_xcom_arg_map.py:194]: I believe that when a task pushes data to XCom, this data can be pulled by another task to dynamically configure the number of task instances to be generated. I believe that the `TaskMap` class leverages this and creates mappings based on the data retireved from XComs, as you can see in the method used `from_task_instance_xcom`, which is used to create a new TaskMap instance from XCom data.

This unittest is essentially verifying the functionality of `from_task_instance_xcom()`, ensuring that it correctly creates a `TaskMap` object from the given `Task_instance` data

Comment (andyjianzhou) [reply tests/models/test_xcom_arg_map.py:194]: but you're right, I'll create one that fails, I think that's important to add

Comment (andyjianzhou) [reply tests/models/test_xcom_arg_map.py:194]: Added!

**Comment by RNHTTR** [comment]
@potiuk a test that seems to be completely unrelated to these changes is failing. Any recommendation here?

**Comment by potiuk** [comment]
> @potiuk a test that seems to be completely unrelated to these changes is failing. Any recommendation here?

I re-run it (next time you can do it as well -> top-right `re-run` button with "re-run failed tests` dropdown.. Mssql server on linux has a tenedency to suddenly stop working and takes a lot of memory (one of the reasons we dumped mssql as backend as it was making our tests way flakier that they should be.

**Review by bbovenzi — APPROVED** [review]
